src/main/scala/com/fulcrumgenomics/commons/async/Async.scala

```diff
@@ -0,0 +1,397 @@
+              outer.throwable.compareAndSet(None, Some(thr))
+              outer.uponException()
+          } finally {
+            outer.uponFinally()
+          }
+        }
+      }
+      val t = new Thread(runnable, name)
+      t.setDaemon(true)
+      t
+    }
+
+    /** Executes the given block of code.  If an [[InterruptedException]] is thrown, throws a [[RuntimeException]]
+      * with the given message.
+      *
+      * @param message the message to use if an [[InterruptedException]] is thrown by the block of code
+      * @param f the block of code to execute
+      */
+    protected def tryAndModifyInterruptedException[T](message: String)(f: => T): T = {
+      try { f } catch {
+        case ie: InterruptedException => throw new RuntimeException(message, ie)
+      }
+    }
+
+    /** Checks to see if an exception has been raised in the asynchronous thread and if so rethrows it as an [[Error]]
+      * or [[RuntimeException]] as appropriate.
+      */
+    protected final def checkAndRethrow(): Unit = {
+      this.throwable.getAndSet(None).foreach { thr =>
+        this.isDone.set(true) // Ensure no further attempts to run
+        thr match {
+          case e: Error             => throw e
+          case ex: RuntimeException => throw ex
+          case t                    => throw new RuntimeException(t)
+        }
+      }
+    }
+  }
+
+
+  /** An asynchronous wrapper for an [[Iterator]].  A separate thread will be created to consume the source iterator.
+    * Will buffer up to [[bufferSize]] elements before the blocking when consuming the source.
+    *
+    * @param source the source iterator to consume
+    * @param bufferSize the maximum number of elements to buffer before the blocking when consuming the source, or None
+    *                   if unbounded.
+    * @tparam T the type of object to consume
+    */
+  class AsyncIterator[T](val source: Iterator[T], bufferSize: Option[Int] = None) extends Iterator[T] with AsyncThread {
+    bufferSize.foreach(size => require(size > 0, s"bufferSize must be greater than zero when given, found $size"))
+
+    private val queue: BlockingQueue[T] = bufferSize match {
+      case Some(size) => new ArrayBlockingQueue[T](size)
+      case None       => new LinkedBlockingDeque[T]()
+    }
+    private val thread = buildThread(getClass.getSimpleName)
+    private val outer  = this
+
+    // Start the thread after creating the queue
+    this.thread.start()
+
+    protected def run(): Unit = outer.source.foreach(outer.queue.put)
+
+    override protected def uponFinally(): Unit = outer.isDone.set(true)
+
+    /** Returns true if there exists more elements, false otherwise */
+    def hasNext(): Boolean = {
+      checkAndRethrow()
+
+      // If the queue is empty but the underlying thread is not done, then there may or may not be new
+      // items that will be put in the qeue.
+      while (this.queue.isEmpty && !this.isDone.get()) {
+        Thread.sleep(50)
```

Comment:
The `Thread.sleep()` with an arbitrary number here bothers me, but I'm struggling to suggest a better solution that's not much more complicated.  

One option would be to implement as:
`!this.queue.isEmpty || ! this.iterator.syncrhonized(this.iterator.isEmpty)`
but that would require synchronizing access to the iterator in the other thread.

Another option would be to have this also be a buffered iterator with a `_next: Option[T]` and have this be:

```scala
def hasNext(): Boolean = {
    while(next.isEmpty && !isDone) _next = Option(queue.poll(50, TimeUnit.Milliseconds))
  _next.isDefined
}
```